src/stream/src/executor/backfill/snapshot_backfill/vnode_stream.rs

```diff
@@ -114,19 +134,29 @@ impl<St: ChangeLogRowStream> VnodeStream<St> {
         })
     }
 
-    #[expect(dead_code)]
+    #[cfg_attr(not(test), expect(dead_code))]
     pub(super) async fn for_vnode_pk_progress(
         &mut self,
         pk_indices: &[usize],
-        mut on_vnode_progress: impl FnMut(VirtualNode, Option<OwnedRow>),
+        mut on_vnode_progress: impl FnMut(VirtualNode, usize, Option<OwnedRow>),
     ) -> StreamExecutorResult<()> {
         assert!(self.data_chunk_builder.is_empty());
-        for vnode in &self.finished_vnode {
-            on_vnode_progress(*vnode, None);
+        for (vnode, row_count) in &self.finished_vnode {
+            on_vnode_progress(*vnode, *row_count, None);
         }
-        for vnode_stream in &mut self.streams {
-            let vnode_stream = vnode_stream.get_mut().expect("should exist");
-            match vnode_stream.as_mut().peek().await {
+        let streams = take(&mut self.streams);
+        // When the `VnodeStream` is polled, the `FuturesUnordered` will generate a special cx to poll the futures.
+        // The pending futures will be stored in a separate linked list and will not be polled until the special cx is awakened
+        // and move the awakened future from the linked list to a ready queue.
+        //
+        // However, here if we use `FuturesUnordered::iter_mut` to in place access the pending futures and call `peek` directly,
```

Comment:
We use iter_mut and peek prior to this pr as well. Does it mean that this issue existed before but we didn't hit it in test?

Reply:
Yes, `peek` is used in `for_vnode_pk_progress`. The method is implemented but not called yet.